File: src/main/java/org/jukeboxmc/block/behavior/BlockDoor.java
package org.jukeboxmc.block.behavior;

import com.nukkitx.nbt.NbtMap;
import com.nukkitx.protocol.bedrock.data.LevelEventType;
import org.jetbrains.annotations.NotNull;
import org.jukeboxmc.block.Block;
import org.jukeboxmc.block.BlockType;
import org.jukeboxmc.block.direction.BlockFace;
import org.jukeboxmc.block.direction.Direction;
import org.jukeboxmc.item.Item;
import org.jukeboxmc.math.Location;
import org.jukeboxmc.math.Vector;
import org.jukeboxmc.player.Player;
import org.jukeboxmc.util.Identifier;
import org.jukeboxmc.world.World;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class BlockDoor extends Block {

    public BlockDoor( Identifier identifier ) {
        super( identifier );
    }

    public BlockDoor( Identifier identifier, NbtMap blockStates ) {
        super( identifier, blockStates );
    }

    @Override
    public boolean placeBlock(@NotNull Player player, @NotNull World world, Vector blockPosition, @NotNull Vector placePosition, Vector clickedPosition, Item itemInHand, BlockFace blockFace ) {
        Block block = world.getBlock( placePosition );
        this.setDirection( Direction.fromAngle( player.getYaw() ) );

        BlockDoor blockAbove = Block.create( this.getType() );
        blockAbove.setLocation( new Location( world, placePosition.add( 0, 1, 0 ) ) );
        blockAbove.setDirection( this.getDirection() );
        blockAbove.setUpperBlock( true );
        blockAbove.setOpen( false );

        Block blockLeft = world.getBlock( placePosition ).getSide( player.getDirection().getLeftDirection() );

        if ( blockLeft.getIdentifier().equals( this.identifier ) ) {
            blockAbove.setDoorHinge( true );
            this.setDoorHinge( true );
        } else {
            blockAbove.setDoorHinge( false );
            this.setDoorHinge( false );
        }

        this.setUpperBlock( false );
        this.setOpen( false );

        world.setBlock( placePosition.add( 0, 1, 0 ), blockAbove, 0 );
        world.setBlock( placePosition, this, 0 );

        /*TODO WATER LOGGING
        if ( block.isWater() ) {
            world.setBlock( placePosition.add( 0, 1, 0 ), block, 1 );
            world.setBlock( placePosition, block, 1 );
        }
         */
        return true;
    }

    @Override
    public boolean interact( Player player, Vector blockPosition, Vector clickedPosition, BlockFace blockFace, Item itemInHand ) {
        this.setOpen( !this.isOpen() );
        this.location.getWorld().sendLevelEvent( this.location, LevelEventType.SOUND_DOOR_OPEN, 0 );
        return true;
    }

    @Override
    public void onBlockBreak(@NotNull Vector breakPosition ) {
        Block block = this.location.getWorld().getBlock( breakPosition, 1 );
        if ( this.isUpperBlock() ) {
            this.location.getWorld().setBlock( this.location.subtract( 0, 1, 0 ), block , 0);
            this.location.getWorld().setBlock( this.location.subtract( 0, 1, 0 ), Block.create( BlockType.AIR ), 1 );
        } else {
            this.location.getWorld().setBlock( this.location.add( 0, 1, 0 ), block, 0 );
            this.location.getWorld().setBlock( this.location.add( 0, 1, 0 ), Block.create( BlockType.AIR ), 1 );
        }
        this.location.getWorld().setBlock( this.location, block, 0 );
        this.location.getWorld().setBlock( this.location, Block.create( BlockType.AIR ), 1 );
    }

    public BlockDoor setOpen( boolean value ) {
        return this.setState( "open_bit", value ? (byte) 1 : (byte) 0 );
    }

    public boolean isOpen() {
        return this.stateExists( "open_bit" ) && this.getByteState( "open_bit" ) == 1;
    }

    public BlockDoor setUpperBlock( boolean value ) {
        return this.setState( "upper_block_bit", value ? (byte) 1 : (byte) 0 );
    }

    public boolean isUpperBlock() {
        return this.stateExists( "upper_block_bit" ) && this.getByteState( "upper_block_bit" ) == 1;
    }

    public BlockDoor setDoorHinge( boolean value ) {
        return this.setState( "door_hinge_bit", value ? (byte) 1 : (byte) 0 );
    }

    public boolean isDoorHinge() {
        return this.stateExists( "door_hinge_bit" ) && this.getByteState( "door_hinge_bit" ) == 1;
    }

    public void setDirection(@NotNull Direction direction ) {
        switch ( direction ) {
            case SOUTH -> this.setState( "direction", 0 );
            case WEST -> this.setState( "direction", 1 );
            case NORTH -> this.setState( "direction", 2 );
            case EAST -> this.setState( "direction", 3 );
        }
    }

    public @NotNull Direction getDirection() {
        int value = this.stateExists( "direction" ) ? this.getIntState( "direction" ) : 0;
        return switch ( value ) {
            case 0 -> Direction.SOUTH;
            case 1 -> Direction.WEST;
            case 2 -> Direction.NORTH;
            default -> Direction.EAST;
        };
    }
}

File: src/main/java/org/jukeboxmc/block/behavior/BlockDoor.kt
package org.jukeboxmc.block.behavior

import com.nukkitx.nbt.NbtMap
import com.nukkitx.protocol.bedrock.data.LevelEventType
import org.jukeboxmc.block.Block
import org.jukeboxmc.block.BlockType
import org.jukeboxmc.block.direction.BlockFace
import org.jukeboxmc.block.direction.Direction
import org.jukeboxmc.item.Item
import org.jukeboxmc.math.Location
import org.jukeboxmc.math.Vector
import org.jukeboxmc.player.Player
import org.jukeboxmc.util.Identifier
import org.jukeboxmc.world.World

/**
 * @author LucGamesYT
 * @version 1.0
 */
class BlockDoor : Block {
    constructor(identifier: Identifier?) : super(identifier)
    constructor(identifier: Identifier?, blockStates: NbtMap?) : super(identifier, blockStates)

    override fun placeBlock(
        player: Player,
        world: World,
        blockPosition: Vector,
        placePosition: Vector,
        clickedPosition: Vector,
        itemInHand: Item,
        blockFace: BlockFace
    ): Boolean {
        val block = world.getBlock(placePosition)
        direction = Direction.Companion.fromAngle(player.yaw)
        val blockAbove: BlockDoor = Block.Companion.create<BlockDoor>(this.type)
        blockAbove.setLocation(Location(world, placePosition.add(0f, 1f, 0f)))
        blockAbove.direction = direction
        blockAbove.setUpperBlock(true)
        blockAbove.setOpen(false)
        val blockLeft = world.getBlock(placePosition).getSide(player.direction.leftDirection)
        if (blockLeft.identifier == identifier) {
            blockAbove.setDoorHinge(true)
            setDoorHinge(true)
        } else {
            blockAbove.setDoorHinge(false)
            setDoorHinge(false)
        }
        setUpperBlock(false)
        setOpen(false)
        world.setBlock(placePosition.add(0f, 1f, 0f), blockAbove, 0)
        world.setBlock(placePosition, this, 0)

        /*TODO WATER LOGGING
        if ( block.isWater() ) {
            world.setBlock( placePosition.add( 0, 1, 0 ), block, 1 );
            world.setBlock( placePosition, block, 1 );
        }
         */return true
    }

    override fun interact(
        player: Player,
        blockPosition: Vector,
        clickedPosition: Vector?,
        blockFace: BlockFace?,
        itemInHand: Item
    ): Boolean {
        setOpen(!isOpen)
        location.world.sendLevelEvent(location, LevelEventType.SOUND_DOOR_OPEN, 0)
        return true
    }

    override fun onBlockBreak(breakPosition: Vector) {
        val block = location.world.getBlock(breakPosition, 1)
        if (isUpperBlock) {
            location.world.setBlock(location!!.subtract(0f, 1f, 0f), block, 0)
            location.world.setBlock(location!!.subtract(0f, 1f, 0f), Block.Companion.create<Block>(BlockType.AIR), 1)
        } else {
            location.world.setBlock(location!!.add(0f, 1f, 0f), block, 0)
            location.world.setBlock(location!!.add(0f, 1f, 0f), Block.Companion.create<Block>(BlockType.AIR), 1)
        }
        location.world.setBlock(location, block, 0)
        location.world.setBlock(location, Block.Companion.create<Block>(BlockType.AIR), 1)
    }

    fun setOpen(value: Boolean): BlockDoor {
        return setState("open_bit", if (value) 1.toByte() else 0.toByte())
    }

    val isOpen: Boolean
        get() = stateExists("open_bit") && getByteState("open_bit").toInt() == 1

    fun setUpperBlock(value: Boolean): BlockDoor {
        return setState("upper_block_bit", if (value) 1.toByte() else 0.toByte())
    }

    val isUpperBlock: Boolean
        get() = stateExists("upper_block_bit") && getByteState("upper_block_bit").toInt() == 1

    fun setDoorHinge(value: Boolean): BlockDoor {
        return setState("door_hinge_bit", if (value) 1.toByte() else 0.toByte())
    }

    val isDoorHinge: Boolean
        get() = stateExists("door_hinge_bit") && getByteState("door_hinge_bit").toInt() == 1
    var direction: Direction
        get() {
            val value = if (stateExists("direction")) getIntState("direction") else 0
            return when (value) {
                0 -> Direction.SOUTH
                1 -> Direction.WEST
                2 -> Direction.NORTH
                else -> Direction.EAST
            }
        }
        set(direction) {
            when (direction) {
                Direction.SOUTH -> setState<Block>("direction", 0)
                Direction.WEST -> setState<Block>("direction", 1)
                Direction.NORTH -> setState<Block>("direction", 2)
                Direction.EAST -> setState<Block>("direction", 3)
            }
        }
}